pkg/ingester/stream.go

```diff
@@ -227,6 +231,18 @@ func (s *stream) Push(
 			chunk = s.cutChunk(ctx)
 		}
 
+		// Check if this this should be rate limited.
+		now := time.Now()
+		if !s.limiter.AllowN(now, s.tenant, len(entries[i].Line)) {
+			validation.DiscardedSamples.WithLabelValues(validation.StreamRateLimit, s.tenant).Add(float64(len(entries) - i))
+			rateLimitesBytes := 0
+			for j := range entries[i:] {
+				rateLimitesBytes += len(entries[j].Line)
+			}
+			validation.DiscardedBytes.WithLabelValues(validation.StreamRateLimit, s.tenant).Add(float64(rateLimitesBytes))
+			return 0, httpgrpc.Errorf(http.StatusTooManyRequests, validation.StreamRateLimitErrorMsg, int(s.limiter.Limit(now, s.tenant)), len(entries)-i, rateLimitesBytes)
```

Comment:
We shouldn't return early here for a few reasons:
1) We'd never run the tailer code.
2) We need to ultimately return the correct `bytesAdded` for partial batches.
3) We want the metrics to be correct for the whole batch, not just one sample.

Instead, try to incorporate these new failures into `failedEntriesWithError`.